code/modules/unit_tests/objectives.dm

```diff
@@ -9,16 +9,37 @@
 			else
 				objectives_that_exist += value
 
+	SStraitor.generate_objectives = FALSE
 	for(var/datum/traitor_objective/objective_typepath as anything in subtypesof(/datum/traitor_objective))
-		if(initial(objective_typepath.abstract_type) == objective_typepath)
+		var/datum/traitor_objective/objective = allocate(objective_typepath)
+		if(objective.abstract_type == objective_typepath)
+			// In this case, we don't want abstract types to define values that should be defined on non-abstract types
+			// Nor do we want abstract types to appear in the pool of traitor objectives.
+			if(objective_typepath in objectives_that_exist)
+				TEST_FAIL("[objective_typepath] is in a traitor category and is an abstract type! Please remove it from the [/datum/traitor_objective_category].")
+			// Since we didn't generate the objective, the rewards are going to be in list form: (min, max)
+			if(!reward_is_zero(objective.progression_reward))
+				TEST_FAIL("[objective_typepath] has set a progression reward as an abstract type! Please define progression rewards on non-abstract types rather than abstract types.")
+			// Since we didn't generate the objective, the rewards are going to be in list form: (min, max)
+			if(!reward_is_zero(objective.telecrystal_reward))
+				TEST_FAIL("[objective_typepath] has set a telecrystal reward as an abstract type! Please define telecrystal rewards on non-abstract types rather than abstract types.")
 			continue
 		if(!(objective_typepath in objectives_that_exist))
 			TEST_FAIL("[objective_typepath] is not in a traitor category and isn't an abstract type! Place it into a [/datum/traitor_objective_category] or remove it from code.")
-		if(initial(objective_typepath.progression_minimum) == null)
+		if(objective.progression_minimum == null)
 			TEST_FAIL("[objective_typepath] has not defined a minimum progression level and isn't an abstract type! Please define the progression minimum variable on the datum")
-		if(!ispath(objective_typepath, /datum/traitor_objective/ultimate) && initial(objective_typepath.progression_reward) == 0 && initial(objective_typepath.telecrystal_reward) == 0)
+		if(objective.needs_reward && reward_is_zero(objective.progression_reward) && reward_is_zero(objective.telecrystal_reward))
 			TEST_FAIL("[objective_typepath] has not set either a progression reward or a telecrystal reward! Please set either a telecrystal or progression reward for this objective.")
 
+/// Returns whether the reward specified (in format (min, max)) is zero or not.
+/datum/unit_test/objectives_category/proc/reward_is_zero(list/reward)
+	return (reward[1] == 0 && reward[2] == 0)
+
+/datum/unit_test/objectives_category/Destroy()
+	SStraitor.generate_objectives = TRUE
+	return ..()
```

Comment:
Is this needed?

Reply:
Which part? The generate_objectives is because it is set to FALSE in the test and we want to guarantee it always resets back. I asked for this

Reply:
yeah, otherwise it will interfere in the traitor.dm test and cause side-effects.
Objectives may not behave differently on creation before and after this test.